### ISSUE TITLE
[Question] How to keep some items after canceling filter or search text returns empty?

### ISSUE THREAD

**Issue by mdtuyen** [davideas/FlexibleAdapter#198]
I custom some AbstractModelItem to make Header same your LayoutItem or ULSItem and put to recyclerview.

When search these item not filter so that them will disappear but I want to show them when I cancel search or clear search.

Please help me make it.
Thanks,

**Comment by mdtuyen** [comment]
Hi @davideas , I saw that when I cancel search filterItems on Item not call, data will restore but not include header so that have no these header.

**Comment by kasterlod** [comment]
Just use it in header item to have effect you want:

@Override
    public boolean filter(String constraint) {
        return false;
    }

**Comment by mdtuyen** [comment]
I understand when item filter() function return false it will not apear (but when constraint is empty this function not call). because adapter.filterItem(data, delay), data not include header.  why header will appear when constaint is empty ?

**Comment by kasterlod** [comment]
data should contains headers and items , new Array, if you write empty constraint or clear search then
 onQueryTextChange(String newText) should call

**Comment by mdtuyen** [comment]
there are some special header need add not same other element as **ULSItem**, **LayoutItem** (NO HeaderItem)  in demo, it not filter and restore when cancel filter.

**Comment by davideas** [comment]
@mdtuyen, I executed some tests. So, if I understood well, you want some items added when the filter is reset, so after a search is cleared out?

Currently the filter is _asynchronously_, it means that, it is executed _after_ the caller method is completed. Indeed if you add items and modify the adapter list, this is executed _before_ the filter. The thread for the filter, with the unfiltered, list will overwrite any modification to the main list!
In the old version of the demo app, I added some new items like **ULSItem**, **LayoutItem** after the search is cleared. Now having the filter _asynchronous_ those items are not present anymore due to the above behavior.

Now some ideas:
- Filter has to remain _asynchronous_.
- I can change the visibility of the methods `postFilter` and `postUpdate` to public so User can override them with `super()` call mandatory. and modify the list after the async execution.

I will restore this feature, I consider this as an enhancement.
You will need to wait until next snapshot version.
